File: Find the number of islands - GFG/find-the-number-of-islands.py
#User function Template for python3
import sys
sys.setrecursionlimit(10**8)
from collections import deque
class Solution:
    def numIslands(self,grid):
        #code here
        n,m = len(grid),len(grid[0])
        visited = [[0]*m for _ in range(n)]
        def dfs(row,col):
            q = deque()
            q.append([row,col])
            visited[row][col] = 1
            while q:
                row,col = q[0]
                q.popleft()
                
                for i in range(-1,2):
                    for j in range(-1,2):
                        nRow = row +  i
                        nCol =  col + j
                        if nRow >= 0 and nRow < n and nCol >= 0 and nCol < m and grid[nRow][nCol] == 1 and not visited[nRow][nCol]:
                            visited[nRow][nCol] = 1
                            q.append([nRow,nCol])
        count = 0
        for i in range(n):
            for j in range(m):
                # print(type(grid[i][j]),visited[i][j])
                if grid[i][j] == 1 and not visited[i][j]:
                    count += 1
                    dfs(i,j)
        return count


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    for _ in range(int(input())):
        n,m=map(int,input().strip().split())
        grid=[]
        for i in range(n):
            grid.append([int(i) for i in input().strip().split()])
        obj=Solution()
        print(obj.numIslands(grid))
# } Driver Code Ends

File: Find the number of islands - GFG/find-the-number-of-islands.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  private:
  void dfs(int row,int col,vector<vector<char>> grid,vector<vector<int>>&visited){
      queue<pair<int,int>> q;
      visited[row][col] = 1;
      q.push({row,col});
      int n = grid.size();
      int m = grid[0].size();
      while(!q.empty()){
          int r = q.front().first;
          int c = q.front().second;
          q.pop();
          
          for(int deltaRow = -1;deltaRow <= 1;deltaRow++){
              for(int deltaCol = -1;deltaCol<=1;deltaCol++){
                  int numRow = r + deltaRow;
                  int numCol = c + deltaCol;
                  if(numRow >= 0 && numRow < n && numCol >= 0 && numCol < m && grid[numRow][numCol]=='1' && !visited[numRow][numCol]){
                      visited[numRow][numCol] = 1;
                      q.push({numRow,numCol});
                  }
              }
          }
      }
  }
  public:
    // Function to find the number of islands.
    int numIslands(vector<vector<char>>& grid) {
        // Code here
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<int>> vis(n,vector<int>(m,0));
        int count = 0;
        for(int i = 0;i < n; i ++){
            for(int j = 0;j < m; j ++ ){
                if(!vis[i][j] and grid[i][j] == '1'){
                    count++;
                    dfs(i,j,grid,vis);
                }
            }
        }
        return count;
    }
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int n, m;
        cin >> n >> m;
        vector<vector<char>> grid(n, vector<char>(m, '#'));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                cin >> grid[i][j];
            }
        }
        Solution obj;
        int ans = obj.numIslands(grid);
        cout << ans << '\n';
    }
    return 0;
}
// } Driver Code Ends